csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedCall.qll

```diff
@@ -0,0 +1,333 @@
+   */
+  abstract predicate hasArguments();
+
+  // TODO: Fix side effects
+  predicate hasReadSideEffect() {
+    any()
+  }
+
+  predicate hasWriteSideEffect() {
+    any()
+  }
+
+  private predicate hasSideEffect() {
+    hasReadSideEffect() or hasWriteSideEffect()
+  }
+
+  override Instruction getPrimaryInstructionForSideEffect(InstructionTag tag) {
+      hasSideEffect() and
+      tag = CallSideEffectTag() and
+      result = getResult()
+  }
+}
+
+/**
+ * IR translation of a direct call to a specific function. Used for both
+ * explicit calls (`TranslatedFunctionCall`) and implicit calls
+ * (`TranslatedAllocatorCall`).
+ */
+abstract class TranslatedDirectCall extends TranslatedCall {
+  override final Instruction getFirstCallTargetInstruction() {
+    result = getInstruction(CallTargetTag())
+  }
+
+  override final Instruction getCallTargetResult() {
+    result = getInstruction(CallTargetTag())
+  }
+
+  override predicate hasInstruction(Opcode opcode, InstructionTag tag,
+      Type resultType, boolean isLValue) {
+    TranslatedCall.super.hasInstruction(opcode, tag, resultType, isLValue) or
+    (
+      tag = CallTargetTag() and
+      opcode instanceof Opcode::FunctionAddress and
+      resultType = expr.getType() and
+      isLValue = true
+    )
+  }
+  
+  override Instruction getInstructionSuccessor(InstructionTag tag,
+      EdgeKind kind) {
+    result = TranslatedCall.super.getInstructionSuccessor(tag, kind) or
+    (
+      tag = CallTargetTag() and
+      kind instanceof GotoEdge and
+      result = getFirstArgumentOrCallInstruction()
+    )
+  }
+}
+
+/**
+ * The IR translation of a call to a function.
+ */
+abstract class TranslatedCallExpr extends TranslatedNonConstantExpr,
+    TranslatedCall {
+  override Call expr;
+
+  override final Type getCallResultType() {
+    result = getResultType()
+  }
+
+  override final predicate hasArguments() {
+    exists(expr.getArgument(0))
+  }
+
+  override final TranslatedExpr getQualifier() {
+  	// TODO: Was getQualifier in the C++ implementation, make sure
+  	//       meaning is kept
+    result = getTranslatedExpr(expr.getChild(-1)) 
+  }
+
+  override final TranslatedExpr getArgument(int index) {
+    result = getTranslatedExpr(expr.getArgument(index))
+  }
+}
+
+/**
+ * Represents the IR translation of a direct function call.
+ */
+class TranslatedFunctionCall extends TranslatedCallExpr, TranslatedDirectCall {
+  override Call expr;
+
+  override Callable getInstructionFunction(InstructionTag tag) {
+    tag = CallTargetTag() and result = expr.getTarget()
+  }
+
+  override predicate hasReadSideEffect() {
+    not expr.getTarget().(SideEffectFunction).neverReadsMemory()
+  }
+
+  override predicate hasWriteSideEffect() {
+    not expr.getTarget().(SideEffectFunction).neverWritesMemory()
+  }
+}
+
+/**
+ * Represents the IR translation of a call to a constructor (can't call destructor explicitly in C#.
+ */
+class TranslatedStructorCall extends TranslatedFunctionCall {
```

Comment:
Might as well call it `TranslatedConstructorCall` then.

Reply:
Was fixed on my current branch.